File: app/src/main/java/de/christinecoenen/code/zapp/utils/view/ClickableViewPager.java
package de.christinecoenen.code.zapp.utils.view;

import android.annotation.SuppressLint;
import android.content.Context;
import androidx.viewpager.widget.ViewPager;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;


public class ClickableViewPager extends ViewPager {

	private View.OnClickListener listener;
	private GestureDetector gestureDetector;

	public ClickableViewPager(Context context) {
		super(context);
		init();
	}

	public ClickableViewPager(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
	}

	private void init() {
		gestureDetector = new GestureDetector(getContext(), new GestureDetector.SimpleOnGestureListener() {
			@Override
			public boolean onSingleTapConfirmed(MotionEvent e) {
				if (listener != null) {
					listener.onClick(null);
				}
				return true;
			}
		});
	}

	@SuppressLint("ClickableViewAccessibility")
	@Override
	public boolean onTouchEvent(MotionEvent event) {
		gestureDetector.onTouchEvent(event);
		return super.onTouchEvent(event);
	}

	@Override
	public void setOnClickListener(View.OnClickListener listener) {
		super.setOnClickListener(listener);
		this.listener = listener;
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/utils/view/ClickableViewPager.kt
package de.christinecoenen.code.zapp.utils.view

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import androidx.viewpager.widget.ViewPager

class ClickableViewPager @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null
) : ViewPager(context, attrs) {

	private var listener: OnClickListener? = null

	private var gestureDetector: GestureDetector = GestureDetector(context, object : SimpleOnGestureListener() {
		override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
			listener?.onClick(null)
			return true
		}
	})

	@SuppressLint("ClickableViewAccessibility")
	override fun onTouchEvent(event: MotionEvent): Boolean {
		gestureDetector.onTouchEvent(event)
		return super.onTouchEvent(event)
	}

	override fun setOnClickListener(listener: OnClickListener?) {
		super.setOnClickListener(listener)
		this.listener = listener
	}
}
